File: java-app/src/main/java/com/example/javaapp/service/UserService.java
package com.example.javaapp.service;

import com.example.javaapp.model.User;
import com.example.javaapp.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * UserService — Business logic layer for user management.
 *
 * LANGUAGE-SPECIFIC FEATURES DEMONSTRATED:
 *
 * 1. No ReentrantLock here — user creation and deletion are simple operations.
 *    The ConcurrentHashMap in UserRepository handles concurrent reads and writes
 *    at the data level. The ReentrantLock is reserved for TaskService where
 *    multi-step operations (check user exists → create task → update user list)
 *    must execute as a single atomic unit.
 *
 * 2. Optional<T> — getUserById() and getUserByUsername() return Optional<User>
 *    instead of null. The caller is forced by the type system to handle the
 *    case where the user does not exist, eliminating NullPointerException.
 *
 * 3. IllegalStateException vs IllegalArgumentException — deleteUser() throws
 *    IllegalStateException (not IllegalArgumentException) when a user still
 *    has tasks. Using the correct exception type communicates the reason for
 *    failure more precisely. IllegalArgument = bad input. IllegalState = the
 *    system is in a state that does not allow this operation.
 */
@Service
public class UserService {

    private final UserRepository userRepository;

    @Autowired
    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }



    public User createUser(String username) {
        if (username == null || username.trim().isEmpty()) {
            throw new IllegalArgumentException("Username cannot be empty");
        }
        if (userRepository.existsByUsername(username.trim())) {
            throw new IllegalArgumentException(
                    "Username already exists: " + username);
        }
        return userRepository.save(new User(username.trim()));
    }

    /**
     * Delete a user only if they have no assigned tasks.
     * Prevents orphaned tasks remaining in the system.
     */
    public boolean deleteUser(String id) {
        Optional<User> userOpt = userRepository.findById(id);
        if (userOpt.isPresent() && userOpt.get().getTaskCount() > 0) {
            throw new IllegalStateException(
                    "Cannot delete user with assigned tasks. " +
                            "Reassign or delete their tasks first.");
        }
        return userRepository.deleteById(id);
    }

    /**
     * Clear all users from the system.
     */
    public void clearAllUsers() {
        userRepository.clear();
    }


    public Optional<User> getUserById(String id) {
        return userRepository.findById(id);
    }

    public Optional<User> getUserByUsername(String username) {
        return userRepository.findByUsername(username);
    }

    public List<User> getAllUsers() {
        return userRepository.findAll();
    }

    public boolean userExists(String id) {
        return userRepository.exists(id);
    }

    public long getUserCount() {
        return userRepository.count();
    }
}

File: js-app/services/UserService.js
import UserRepository from '../repositories/UserRepository.js';

/*
 * UserService — Business logic for user management.
 *
 * LANGUAGE-SPECIFIC FEATURES DEMONSTRATED:
 *
 * 1. async/await — every method is async because repository operations
 *    involve file I/O. The caller awaits the result without blocking
 *    the event loop. Other requests can be handled while I/O completes.
 *
 * 2. No locking — unlike Java's UserService which uses ReentrantLock,
 *    no synchronization is needed here. Node.js is single-threaded:
 *    only one piece of JavaScript runs at a time, so Map operations
 *    are never truly concurrent.
 *
 * 3. Error throwing with plain Error objects — JavaScript uses
 *    throw new Error('message') where Java uses typed exceptions
 *    (IllegalArgumentException, IllegalStateException). JS has no
 *    checked exceptions — the caller is NOT forced to handle errors
 *    by the compiler. This is a key language difference.
 *
 * 4. Truthy/falsy checks — if (!username) catches both null and
 *    empty string. Java requires explicit null checks and .isEmpty().
 */
class UserService {

    constructor() {
        this.userRepository = new UserRepository();
    }

    async createUser(username) {
        if (!username || username.trim() === '') {
            throw new Error('Username cannot be empty');
        }
        const trimmed = username.trim();
        if (await this.userRepository.existsByUsername(trimmed)) {
            throw new Error(`Username already exists: ${trimmed}`);
        }
        // Import User lazily to avoid circular dependency issues
        const { default: User } = await import('../models/User.js');
        const user = new User(trimmed);
        return await this.userRepository.save(user);
    }

    async deleteUser(id) {
        const user = await this.userRepository.findById(id);
        if (user && user.getTaskCount() > 0) {
            throw new Error('Cannot delete user with assigned tasks. Remove their tasks first.');
        }
        return await this.userRepository.deleteById(id);
    }


    async getUserById(id) {
        return await this.userRepository.findById(id);
    }

    async getUserByUsername(username) {
        return await this.userRepository.findByUsername(username);
    }

    async getAllUsers() {
        return await this.userRepository.findAll();
    }

    async userExists(id) {
        return await this.userRepository.exists(id);
    }

    async getUserCount() {
        return await this.userRepository.count();
    }

    async clearAllUsers() {
        return await this.userRepository.clear();
    }
}

export default UserService;